_posts/2019-09-12-API-first-development-with-OpenAPI-or-Swagger.md

```diff
@@ -0,0 +1,549 @@
+tags: [API, cloud]
+category: Cloud
+comments: true
+---
+
+# Table of content
+
+1.[API first development - Why, how and what](#api-first-development---why-how-and-what)  
+2.[API design: an example](#api-design-an-example)  
+3.[API first development with OpenAPI/Swagger](#api-first-development-with-openapiswagger)  
+4.[Hosting your visualisations](#hosting-your-visualisations)  
+5.[Integrating with Postman](#integrating-with-postman)  
+6.[OpenAPI Generator: generate API compliant code](#openapi-generator-generate-api-compliant-code)  
+7.[Integrating with your build process: maven or gradle plugin](#integrating-swaggeropenapi-with-your-build-process-maven-or-gradle-plugin)  
+8.[Serverless on AWS: OpenAPI, API Gateway Lambda and SAM](#serverless-on-aws-openapi-api-gateway-lambda-and-sam)  
+9.[Springfox](#code-first-with-springfox)   
+10.[Conclusion](#conclusion)
+
+
+
+# Intro
+* I'll start of with a lecture about API first development and it's advantages.([Jump to section](#api-first-development---why-how-and-what))
+* We will discover how to visualize our API specs. ([Jump to section](#api-first-development-with-openapiswagger))  
+* Generate a postman collection from your OpenAPI definition. ([Jump to section](#integrating-with-postman))   
+* Next I'll show you how to generate code that is completely compliant with your API specs. ([Jump to section](#openapi-generator-generate-api-compliant-code))
+* We then dive into how to integrate this in your build process by using maven or gradle plugins. ([Jump to section](#integrating-with-your-build-process-maven-or-gradle-plugin))
+* To finish it up I'll demonstrate how to use it in a cloud native serverless product with `AWS SAM` and `AWS Lambda`. ([Jump to section](#serverless-on-aws-openapi-api-gateway-lambda-and-sam))
+* Wrapping it up with a conclusion.
+
+# API first development - Why, how and what
+> Great communication is key to great software engineering.
+
+That also goes for applications and services.
+If your applications cannot communicate properly, you'll never be able to expose the functionality that is key to a good product.
+
+
+We also see the following trends in software engineering:
+* **Shift towards the cloud.**  
+Instead of big monolitic applications we are building lot's of smaller (micro)services.
+All communication between those services goes through the API.
+
+* **Multiple frontend applications use the same backend.**  
+Often these applications are created by separate teams.
+
+* **API carry business value.**  
+Yes, there is money in your API.
+An API exposes the functionality of your product.
+A good API allows user to integrate with your product with ease.
+Thus making your product a great choice.
+
+In all of the cases a above there is value (even money!) in good communication between services.
+And that's why you should practice API first development.
+Put your communication first!
+
+> API first development makes a clear API definition the first priority
```

Comment:
The first priority in your API first development story is a clear API definition